### ISSUE TITLE
key is not defined

### ISSUE THREAD

**Issue by fanmanqian** [JusperLee/Conv-TasNet#13]
During the inference,key is not defined .what is the value of key?
![Uploading QQ图片20200731110653.jpg…]()

**Comment by fanmanqian** [comment]
解决了。

**Comment by xiao-yu0** [comment]
> 解决了。
可以请问一下您怎么解决的，我也遇到了这个问题

**Comment by JusperLee** [comment]
You can refer to the following code and add a key option in dataset.py.
```python
import torch
from torch.utils.data import DataLoader, Dataset
import torch.nn.functional as F
import random
from utils import handle_scp
import numpy as np
import soundfile as sf
import torchaudio


def read_wav(fname, return_rate=False):
    '''
         Read wavfile using Pytorch audio
         input:
               fname: wav file path
               return_rate: Whether to return the sampling rate
         output:
                src: output tensor of size C x L 
                     L is the number of audio frames 
                     C is the number of channels. 
                sr: sample rate
    '''
    src, sr = sf.read(fname, start=0,
                       stop=None, dtype='float32')
    src = torch.from_numpy(src)
    if return_rate:
        return src.squeeze(), sr
    else:
        return src.squeeze()


def make_dataloader(is_train=True,
                    data_kwargs=None,
                    num_workers=4,
                    chunk_size=32000,
                    batch_size=16):
    dataset = Datasets(**data_kwargs, chunk_size=chunk_size)
    return DataLoader(dataset,
                      batch_size=batch_size,
                      num_workers=num_workers,
                      shuffle=is_train,
                      drop_last=True)


class Datasets(Dataset):
    '''
       Load audio data
       mix_scp: file path of mix audio (type: str)
       ref_scp: file path of ground truth audio (type: list[spk1,spk2])
    '''

    def __init__(self, mix_scp=None, ref_scp=None, sr=8000, chunk_size=32000, test=False):
        super(Datasets, self).__init__()
        self.mix_audio = handle_scp(mix_scp)
        self.ref_audio = [handle_scp(r) for r in ref_scp]
        self.sr = sr
        self.chunk_size = chunk_size
        self.test = test
        self.key = list(self.mix_audio.keys())
        self.key_tmp = self.key.copy()
        if self.test != True:
            for k in self.key_tmp:
                mix = read_wav(self.mix_audio[k])
                if mix.shape[-1] < 32000:
                    self.key.remove(k)

    def __len__(self):
        return len(self.key)

    def __getitem__(self, index):
        index = self.key[index]
        mix = read_wav(self.mix_audio[index])
        ref = [read_wav(r[index]) for r in self.ref_audio]
        if mix.shape[-1] == self.chunk_size or self.test:
            rand_start = 0
        else:
            rand_start = np.random.randint(0, mix.shape[-1] - self.chunk_size)
        if self.test:
            stop = None
        else:
            stop = rand_start + self.chunk_size
        mix, _ = sf.read(self.mix_audio[index], start=rand_start,
                         stop=stop, dtype='float32')
        for i in range(len(ref)):
            ref[i], _ = sf.read(self.ref_audio[i][index], start=rand_start,
                                stop=stop, dtype='float32')
        return {
            'mix': torch.from_numpy(mix),
            'ref': [torch.from_numpy(r) for r in ref],
            'key': index
        }

if __name__ == "__main__":
    datasets = Datasets('/home/likai/data1/create_scp/tr_mix.scp',
                        ['/home/likai/data1/create_scp/tr_s1.scp', '/home/likai/data1/create_scp/tr_s2.scp'], test=False)
    print(len(datasets))
```

**Comment by JusperLee** [comment]
@xiao-yu0

**Comment by xiao-yu0** [comment]
您好，请问您有WSJ0数据集吗，可不可以分享一下，我只需要部分来测试一下，我运行了你的代码还是没有运行出来，不知道是不是文件路径出了问题，所以想看一下这个数据集的文件结构。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "JusperLee/Conv-TasNet"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年9月23日(星期三) 晚上11:09
收件人:&nbsp;"JusperLee/Conv-TasNet"<Conv-TasNet@noreply.github.com&gt;;
抄送:&nbsp;"1092162049"<1092162049@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [JusperLee/Conv-TasNet] key is not defined (#13)





 
You can refer to the following code and add a key option in dataset.py.
 import torch from torch.utils.data import DataLoader, Dataset import torch.nn.functional as F import random from utils import handle_scp import numpy as np import soundfile as sf import torchaudio def read_wav(fname, return_rate=False):     '''          Read wavfile using Pytorch audio          input:                fname: wav file path                return_rate: Whether to return the sampling rate          output:                 src: output tensor of size C x L                       L is the number of audio frames                       C is the number of channels.                  sr: sample rate     '''     src, sr = sf.read(fname, start=0,                        stop=None, dtype='float32')     src = torch.from_numpy(src)     if return_rate:         return src.squeeze(), sr     else:         return src.squeeze() def make_dataloader(is_train=True,                     data_kwargs=None,                     num_workers=4,                     chunk_size=32000,                     batch_size=16):     dataset = Datasets(**data_kwargs, chunk_size=chunk_size)     return DataLoader(dataset,                       batch_size=batch_size,                       num_workers=num_workers,                       shuffle=is_train,                       drop_last=True) class Datasets(Dataset):     '''        Load audio data        mix_scp: file path of mix audio (type: str)        ref_scp: file path of ground truth audio (type: list[spk1,spk2])     '''     def __init__(self, mix_scp=None, ref_scp=None, sr=8000, chunk_size=32000, test=False):         super(Datasets, self).__init__()         self.mix_audio = handle_scp(mix_scp)         self.ref_audio = [handle_scp(r) for r in ref_scp]         self.sr = sr         self.chunk_size = chunk_size         self.test = test         self.key = list(self.mix_audio.keys())         self.key_tmp = self.key.copy()         if self.test != True:             for k in self.key_tmp:                 mix = read_wav(self.mix_audio[k])                 if mix.shape[-1] < 32000:                     self.key.remove(k)     def __len__(self):         return len(self.key)     def __getitem__(self, index):         index = self.key[index]         mix = read_wav(self.mix_audio[index])         ref = [read_wav(r[index]) for r in self.ref_audio]         if mix.shape[-1] == self.chunk_size or self.test:             rand_start = 0         else:             rand_start = np.random.randint(0, mix.shape[-1] - self.chunk_size)         if self.test:             stop = None         else:             stop = rand_start + self.chunk_size         mix, _ = sf.read(self.mix_audio[index], start=rand_start,                          stop=stop, dtype='float32')         for i in range(len(ref)):             ref[i], _ = sf.read(self.ref_audio[i][index], start=rand_start,                                 stop=stop, dtype='float32')         return {             'mix': torch.from_numpy(mix),             'ref': [torch.from_numpy(r) for r in ref],             'key': index         } if __name__ == "__main__":     datasets = Datasets('/home/likai/data1/create_scp/tr_mix.scp',                         ['/home/likai/data1/create_scp/tr_s1.scp', '/home/likai/data1/create_scp/tr_s2.scp'], test=False)     print(len(datasets))
